packages/editor/src/lib/primitives/geometry/Geometry2d.ts

```diff
@@ -73,10 +109,30 @@ export abstract class Geometry2d {
 		return this.isClosed && this.isFilled && pointInPolygon(nearest, this.vertices) ? -dist : dist
 	}
 
-	hitTestLineSegment(A: Vec, B: Vec, distance = 0): boolean {
-		return this.distanceToLineSegment(A, B) <= distance
+	hitTestLineSegment(A: Vec, B: Vec, distance = 0, filters?: Geometry2dFilters): boolean {
+		return this.distanceToLineSegment(A, B, filters) <= distance
+	}
+
+	*intersectLineSegment(A: VecLike, B: VecLike, filters?: Geometry2dFilters) {
+		if (this.isExcludedByFilter(filters)) return
+
+		const intersections = this.isClosed
+			? intersectLineSegmentPolygon(A, B, this.vertices)
+			: intersectLineSegmentPolyline(A, B, this.vertices)
+
+		if (intersections) yield* intersections
+	}
+
+	*intersectCircle(center: VecLike, radius: number, filters?: Geometry2dFilters) {
+		if (this.isExcludedByFilter(filters)) return
+		const intersections = this.isClosed
+			? intersectCirclePolygon(center, radius, this.vertices)
+			: intersectCirclePolyline(center, radius, this.vertices)
+
+		if (intersections) yield* intersections
```

Comment:
why use a generator here? I'm a bit worried about perf vs just returning the actual array. Geometry operations are liable to be called frequently in hot loops and on code paths that block UI updates, so we should probably avoid this extra allocation and indirection if it's not really needed

Reply:
My main thinking here was groups, where you might have intersections from several pieces of geometry that you'd otherwise need to merge, but i guess that's maybe less allocations overall than using a generator (?)